clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -350,6 +350,10 @@ private KafkaChannel buildAndAttachKafkaChannel(SocketChannel socketChannel, Str
             // Ideally, these resources are closed by the KafkaChannel but if the KafkaChannel is not created to an
             // error, this builder should close the resources it has created instead.
             Utils.closeQuietly(metadataRegistry, "metadataRegistry");
+            // If an Authentication Failure is encountered, throw that error directly.
+            if (e instanceof AuthenticationException) {
+                throw e;
```

Comment:
There are callers, e.g. NetworkClient, which doesn't expect an AuthenticationException while connecting, which could result in inconsistent states in the client. We need to update those and add tests for those too. I would also suggest adding an integration test with a Producer that handles the exception.

Reply:
@rajinisivaram , i made the changes as you suggested (added UT's as well). Regarding the Integration test, I tried adding them in `kafka.api.SaslClientsWithInvalidCredentialsTest` but I am not able to replicate the exact scenario. In the producer test `testProducerWithAuthenticationFailure` in that class, sending the record fails till we don't setup the client credentials. I have been wanting to add a test where the credentials expire and the producer also fails immediately(instead of the current behaviour of blocking for max.block.ms). Is there any existing test which mimics this ?

Reply:
Don't know of an existing test that verifies the case that the PR is trying to fix. So you may need to write your own test for the specific scenario.